### PR TITLE
ConfigurationGathering: Use `$EDITOR` if set

### PR DESCRIPTION
**Pull request by adtac**
### Checklist

- [x] I have rebased properly. Please see [our tutorial on rebasing](http://coala.readthedocs.io/en/latest/Developers/Git_Basics.html#rebasing).
- [x] I have gone through the [commit guidelines](http://coala.readthedocs.io/en/latest/Developers/Writing_Good_Commits.html) and I've followed them.
- [x] I have followed the [guidelines for the commit shortlog](http://docs.coala.io/en/latest/Developers/Writing_Good_Commits.html#shortlog).
- [x] I have followed the [guidelines for the commit body](http://docs.coala.io/en/latest/Developers/Writing_Good_Commits.html#commit-body).
- [x] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://docs.coala.io/en/latest/Developers/Writing_Good_Commits.html#issue-reference).
- [x] I have [run all the tests](http://coala.readthedocs.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

### Reviewers

If the `$EDITOR` environment variable set, use it.

Closes https://github.com/coala/coala/issues/2299

### CODE REVIEW

File: tests/settings/ConfigurationGatheringTest.py [hunk 1]
```diff
@@ -276,3 +279,31 @@ def test_no_config(self):
...
+        with change_directory(child_dir):
+            os.environ["EDITOR"] = "vim"
+            sections, local_bears, global_bears, targets = gather_configuration(
+                lambda *args: true,
+                self.log_printer,
+                arg_list=[])
+            # ``$EDITOR`` has preference over whatever is set in setttings
+            # (which can be from a ``.coafile`` or ``.coarc``) because when
+            # the user runs ``EDITOR=vim ./coala``, it should respect that.
+            # Also it seems only natural to override emacs with vim :P
```
Comment (adtac) [inline tests/settings/ConfigurationGatheringTest.py]: This is a joke :P
If you feel someone might be offended, I'll remove it :)

**Comment by jayvdb** [comment]
I think the order should be CLI arg (if any), `.coarc`, `EDITOR`, `.coafile`, others... ?

File: coalib/settings/ConfigurationGathering.py [hunk 2]
```diff
@@ -179,6 +179,10 @@ def load_configuration(arg_list, log_printer, arg_parser=None):
 
         sections = merge_section_dicts(sections, cli_sections)
 
+    if "EDITOR" in os.environ and "editor" not in sections["default"]:
+        sections["default"].add_or_create_setting(
+            Setting("editor", os.environ["EDITOR"]))
```
Comment (sils) [inline coalib/settings/ConfigurationGathering.py]: I think this is the wrong place to do this in, rather do it right in the OpenEditorAction? The EDITOR shouldn't be written to the sections as we may also decide to store it back to file later

File: tests/output/ConsoleInteractionTest.py [hunk 3]
```diff
@@ -137,10 +137,13 @@ def setUp(self):
             spaces="•",
             tabs=True,
             tabsize=SpacingHelper.DEFAULT_TAB_WIDTH))
+        self.old_environ = os.environ
+        os.environ = {}
 
     def tearDown(self):
         OpenEditorAction.is_applicable = self.old_open_editor_applicable
         ApplyPatchAction.is_applicable = self.old_apply_patch_applicable
+        os.environ = self.old_environ
```
Comment (underyx) [inline tests/output/ConsoleInteractionTest.py]: I'd just use pytest's [`monkeypatch.setenv`](http://doc.pytest.org/en/latest/monkeypatch.html#_pytest.monkeypatch.MonkeyPatch.setenv) instead.

File: tests/results/result_actions/OpenEditorActionTest.py [hunk 4]
```diff
@@ -2,7 +2,10 @@
 import subprocess
 import tempfile
 import unittest
+from importlib import reload
 
+import coalib.results.result_actions.OpenEditorAction
```
Comment (adtac) [inline tests/results/result_actions/OpenEditorActionTest.py]: Required since we have to reload it in the future

**Comment by sils** [comment]
fixes -> closes hm?

**Comment by sils** [comment]
also rebase

**Comment by sils** [comment]
fixed it up for you on https://github.com/coala/coala/pull/3047

**Pull request closed without merging**